service/history/failoverCheck.go

```diff
@@ -61,3 +64,77 @@ func verifyTimerTaskVersion(shard ShardContext, domainID string, version int64,
 	}
 	return true, nil
 }
+
+// load mutable state, if mutable state's next event ID <= task ID, will attempt to refresh
+// if still mutable state's next event ID <= task ID, will return nil, nil
+func loadMutableStateForTransferTask(context *workflowExecutionContext, transferTask *persistence.TransferTaskInfo, metricsClient metrics.Client, logger bark.Logger) (*mutableStateBuilder, error) {
+	msBuilder, err := context.loadWorkflowExecution()
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); ok {
+			// this could happen if this is a duplicate processing of the task, and the execution has already completed.
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	// check to see if cache needs to be refreshed as we could potentially have stale workflow execution
+	// the exception is decision consistently fail
+	// there will be no event generated, thus making the decision schedule ID == next event ID
+	isDecisionRetry := transferTask.TaskType == persistence.TransferTaskTypeDecisionTask &&
+		msBuilder.executionInfo.DecisionScheduleID == transferTask.ScheduleID &&
+		msBuilder.executionInfo.DecisionAttempt > 0
+
+	if transferTask.ScheduleID >= msBuilder.GetNextEventID() && !isDecisionRetry {
+		metricsClient.IncCounter(metrics.TimerQueueProcessorScope, metrics.StaleMutableStateCounter)
```

Comment:
Probably you want to use a different metric scope here.

Reply:
fixed